File: day15/solution.py
#!/usr/bin/env python3

# DEBUG = True
DEBUG = False
# FILENAME = 'example.txt'
# FILENAME = 'example2.txt'
# FILENAME = 'example3.txt'
FILENAME = 'input.txt'
# print(f'{FILENAME=}')

with open(FILENAME, 'r') as f:
  text = f.read()
# if DEBUG: print(f'{text=}')
sections = text.split('\n\n')
# if DEBUG: print(f'{sections=}')


def part2(sections: list[str]):
  map_data = []
  for line in sections[0].splitlines():
    map_line = []
    for char in line:
      match char:
        case 'O':
          map_line.extend(['[', ']'])
        case '@':
          map_line.extend(['@', '.'])
        case _:
          map_line.extend([char, char])
    map_data.append(map_line)

  moves = [x for x in sections[1] if x.strip()]
  # if DEBUG: print(f'{moves=}')
  len_rows = len(map_data)
  len_cols = len(map_data[0])

  def find_start():
    for r in range(len_rows):
      for c in range(len_cols):
        if map_data[r][c] == '@':
          # map_data[r][c] = '.'
          return (r, c)
    assert False, 'start not found'

  def print_map():
    for r in map_data:
      print(''.join(r))
    print()

  pos_row, pos_col = find_start()
  if DEBUG: print(f'{pos_row=} {pos_col=}')

  if DEBUG: print_map()

  for move_ix, move in enumerate(moves):
    move_row = -1 if move == '^' else 1 if move == 'v' else 0
    move_col = -1 if move == '<' else 1 if move == '>' else 0
    boxes: list[tuple[int, int]] = []
    test_cols = [pos_col]
    wall = False
    if DEBUG: print(f'{move_ix=} BEGIN {move=} {move_row=} {move_col=} {pos_row=} {pos_col=}')
    for i in range(1, max(len_rows, len_cols)):
      test_row = pos_row + move_row * i
      if move_col != 0:
        test_cols = [x + move_col for x in test_cols]

      new_test_cols = []
      for test_col in test_cols.copy():
        if DEBUG: print(f'{test_col=} {test_cols=}')
        test_char = map_data[test_row][test_col]
        match test_char:
          case '.':
            test_cols.remove(test_col)
            if len(test_cols) == 0:
              # move
              for box_row, box_col in reversed(boxes):
                box_row_to = box_row + move_row
                box_col_to = box_col + move_col
                if DEBUG: print(f'move {box_row=} {box_col=} to {box_row_to} {box_col_to=}')
                map_data[box_row_to][box_col_to] = map_data[box_row][box_col]
                map_data[box_row][box_col] = '.'
              map_data[pos_row][pos_col] = '.'
              pos_row += move_row
              pos_col += move_col
              map_data[pos_row][pos_col] = '@'
              break

          case '[' | ']':

            if DEBUG: print(f'box {test_char} at {test_row=} {test_col=}')
            test_box = (test_row, test_col)
            if test_box not in boxes:
              boxes.append(test_box)
            if test_col not in new_test_cols:
              new_test_cols.append(test_col)
            if move_row != 0:
              new_test_col = test_col + (1 if test_char == '[' else -1)
              new_test_box = (test_row, new_test_col)
              if new_test_box not in boxes:
                boxes.append(new_test_box)
              if new_test_col not in new_test_cols:
                new_test_cols.append(new_test_col)
              if DEBUG: print(f'{new_test_col=} {test_cols=}')
            if DEBUG: print(f'{boxes=} {new_test_cols}')
          case '#':
            if DEBUG: print(f'wall at {test_row=} {test_col=}')
            wall = True
            break

      test_cols = new_test_cols
      if wall or len(new_test_cols) == 0: break

    if DEBUG:
      print(f'{move_ix} END {move=} {pos_row=} {pos_col=}')
      print_map()

  gps_sum = 0
  for ix_row in range(len_rows):
    for ix_col in range(len_cols):
      if map_data[ix_row][ix_col] == '[':
        gps = 100 * ix_row + ix_col
        if DEBUG: print(f'{ix_row=} {ix_col=} {gps=}')
        gps_sum += gps
  print(f'part 2: {gps_sum}')


def part1(sections: list[str]):
  map_data = [list(x) for x in sections[0].splitlines()]
  moves = [x for x in sections[1] if x.strip()]
  # if DEBUG: print(f'{moves=}')
  len_rows = len(map_data)
  len_cols = len(map_data[0])

  def find_start():
    for r in range(len_rows):
      for c in range(len_cols):
        if map_data[r][c] == '@':
          # map_data[r][c] = '.'
          return (r, c)
    assert False, 'start not found'

  def print_map():
    for r in map_data:
      print(''.join(r))
    print()

  pos_row, pos_col = find_start()
  if DEBUG: print(f'{pos_row=} {pos_col=}')

  if DEBUG: print_map()

  for move_ix, move in enumerate(moves):
    move_row = -1 if move == '^' else 1 if move == 'v' else 0
    move_col = -1 if move == '<' else 1 if move == '>' else 0
    box_row = -1
    box_col = -1
    if DEBUG: print(f'{move_ix=} BEGIN {move=} {move_row=} {move_col=} {pos_row=} {pos_col=}')
    for i in range(1, max(len_rows, len_cols)):
      test_row = pos_row + move_row * i
      test_col = pos_col + move_col * i
      if DEBUG: print(f'{test_row=} {test_col=}')
      match map_data[test_row][test_col]:
        case '.':
          if DEBUG: print(f'empty at {test_row=} {test_col=}')
          # empty
          if box_row != -1:
            # move first encountered box to here
            if DEBUG: print(f'move box from {box_row=} {box_col=} to {test_row=} {test_col=}')
            map_data[test_row][test_col] = 'O'
            map_data[box_row][box_col] = '.'
          # increment pos by move
          map_data[pos_row][pos_col] = '.'
          pos_row += move_row
          pos_col += move_col
          map_data[pos_row][pos_col] = '@'
          break
        case 'O':
          # set box if first encountered
          if DEBUG: print(f'box at {test_row=} {test_col=}')
          if box_row == -1:
            box_row = test_row
            box_col = test_col
        case '#':
          # wall - no move
          if DEBUG: print(f'wall at {test_row=} {test_col=}')
          break
    if DEBUG:
      print(f'{move_ix} END {move=} {pos_row=} {pos_col=}')
      print_map()

  gps_sum = 0
  for ix_row in range(len_rows):
    for ix_col in range(len_cols):
      if map_data[ix_row][ix_col] == 'O':
        gps = 100 * ix_row + ix_col
        if DEBUG: print(f'{ix_row=} {ix_col=} {gps=}')
        gps_sum += gps
  print(f'part 1: {gps_sum}')


part1(sections)
part2(sections)

File: day15/solution.ts
#!/usr/bin/env -S deno --allow-read
// deno-lint-ignore-file no-case-declarations

// const DEBUG = true;
const DEBUG = false;
// const FILENAME = 'example.txt';
// const FILENAME = 'example2.txt';
// const FILENAME = 'example3.txt';
// const FILENAME = 'example4.txt';
const FILENAME = 'input.txt';
// console.log(FILENAME);

const text = await Deno.readTextFile(FILENAME);
// if (DEBUG) console.debug({ text });
const blocks = text.split('\n\n');
// if (DEBUG) console.debug({ blocks });
const moves = blocks[1].split('').filter((char) => char.trim() != '');
// if (DEBUG) console.debug({ moves });

const printMap = (mapData: string[][], rowFrom: number = 0, rowTo: number = -1, colFrom: number = 0, colTo: number = -1) => {
  let rowIx = rowFrom;
  for (const row of mapData.slice(rowFrom, rowTo == -1 ? undefined : rowTo)) {
    const rowIxStr = String(rowIx).padStart(3, '0');
    console.log(`${rowIxStr} ${row.slice(colFrom, colTo == -1 ? undefined : colTo).join('')}`);
    rowIx++;
  }
};

const getMapParams = (mapData: string[][]) => {
  const rowCount = mapData.length;
  const colCount = mapData[0].length;
  for (let startRowIx = 0; startRowIx < rowCount; startRowIx++) {
    for (let startColIx = 0; startColIx < colCount; startColIx++) {
      if (mapData[startRowIx][startColIx] === '@') {
        return { rowCount, colCount, startRowIx, startColIx };
      }
    }
  }
  throw new Error('could not find start position');
};

const getOffset = (move: string) => {
  switch (move) {
    case '^':
      return { offsetRow: -1, offsetCol: 0 };
    case '>':
      return { offsetRow: 0, offsetCol: 1 };
    case 'v':
      return { offsetRow: 1, offsetCol: 0 };
    case '<':
      return { offsetRow: 0, offsetCol: -1 };
  }
  throw new Error(`unknown move ${move}`);
};

const part1 = (mapText: string, moves: string[]) => {
  const mapData = mapText
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => line.split(''));
  if (DEBUG) printMap(mapData);
  const { rowCount, colCount, startRowIx, startColIx } = getMapParams(mapData);
  let row = startRowIx;
  let col = startColIx;
  if (DEBUG) console.debug({ rowCount, colCount, row, col });
  for (const move of moves) {
    const { offsetRow, offsetCol } = getOffset(move);
    const newRow = row + offsetRow;
    const newCol = col + offsetCol;
    if (DEBUG) console.debug({ row, col, move, offsetRow, offsetCol, newRow, newCol });
    switch (mapData[newRow][newCol]) {
      case '.':
        mapData[row][col] = '.';
        row = newRow;
        col = newCol;
        mapData[row][col] = '@';
        break;
      case '#':
        break;
      case 'O':
        let boxRow = newRow;
        let boxCol = newCol;
        while (mapData[boxRow][boxCol] == 'O') {
          boxRow += offsetRow;
          boxCol += offsetCol;
        }
        if (mapData[boxRow][boxCol] == '.') {
          mapData[row][col] = '.';
          mapData[boxRow][boxCol] = 'O';
          row = newRow;
          col = newCol;
          mapData[row][col] = '@';
        }
        break;
      default:
        throw new Error(`invalid map char ${mapData[newRow][newCol]} at ${newRow},${newCol}`);
    }
    if (DEBUG) printMap(mapData);
  }
  const gpsTotal = mapData.reduce(
    (rowAcc, row, rowIx) => (rowAcc += row.reduce((colAcc, char, colIx) => (colAcc += char == 'O' ? rowIx * 100 + colIx : 0), 0)),
    0
  );
  console.log('part 1:', gpsTotal);
};

const pushBoxCoords = (boxCoords: number[][], newCoords: number[][]) => {
  //TODO: this is very inneficient. map of a coord class with a hash function would be faster
  for (const newCoord of newCoords) {
    let duplicate = false;
    for (const boxCoord of boxCoords) {
      if (boxCoord[0] == newCoord[0] && boxCoord[1] == newCoord[1]) {
        duplicate = true;
        break;
      }
    }
    if (!duplicate) {
      boxCoords.push(newCoord);
    }
  }
  return boxCoords;
};

const walkVertBoxes = (
  mapData: string[][],
  row: number,
  col: number,
  offset: number
): { boxCoords: number[][]; success: boolean } => {
  // recursive horrors
  // i think duplication is unavoidable using this approach
  let boxCoords: number[][] = [];
  let success = true;
  switch (mapData[row][col]) {
    case '#':
      success = false;
      break;
    case '[':
      const { boxCoords: llBoxCoords, success: llSuccess } = walkVertBoxes(mapData, row + offset, col, offset);
      const { boxCoords: lrBoxCoords, success: lrSuccess } = walkVertBoxes(mapData, row + offset, col + 1, offset);
      success = success && llSuccess && lrSuccess;
      boxCoords = pushBoxCoords(boxCoords, [[row, col], [row, col + 1], ...llBoxCoords, ...lrBoxCoords]);
      break;
    case ']':
      const { boxCoords: rlBoxCoords, success: rlSuccess } = walkVertBoxes(mapData, row + offset, col - 1, offset);
      const { boxCoords: rrBoxCoords, success: rrSuccess } = walkVertBoxes(mapData, row + offset, col, offset);
      success = success && rlSuccess && rrSuccess;
      boxCoords = pushBoxCoords(boxCoords, [[row, col - 1], [row, col], ...rlBoxCoords, ...rrBoxCoords]);
      break;
    case '.':
      success = true;
      break;
  }
  for (const outer of boxCoords) {
    let count = 0;
    for (const inner of boxCoords) {
      if (inner[0] == outer[0] && inner[1] == outer[1]) count++;
    }
    if (count > 1) throw new Error(`duplicate boxCoords ${outer} count=${count}`);
  }
  return { boxCoords, success };
};

const part2 = (mapText: string, moves: string[]) => {
  const mapData = mapText
    .split('\n')
    .filter((line) => line.trim())
    .map((line) =>
      line
        .split('')
        .map((char) => {
          switch (char) {
            case '@':
              return ['@', '.'];
            case 'O':
              return ['[', ']'];
            default:
              return [char, char];
          }
        })
        .flat(1)
    );
  if (DEBUG) printMap(mapData);
  const { rowCount, colCount, startRowIx, startColIx } = getMapParams(mapData);
  let row = startRowIx;
  let col = startColIx;
  if (DEBUG) console.debug({ rowCount, colCount, row, col });
  let moveId = 0;
  for (const move of moves) {
    moveId++;
    const { offsetRow, offsetCol } = getOffset(move);
    const newRow = row + offsetRow;
    const newCol = col + offsetCol;
    // if (DEBUG) console.debug({ moveId, row, col, move, offsetRow, offsetCol, newRow, newCol });
    const charAt = mapData[newRow][newCol];
    switch (charAt) {
      case '.':
        mapData[row][col] = '.';
        row = newRow;
        col = newCol;
        mapData[row][col] = '@';
        break;
      case '#':
        break;
      case '[':
      case ']':
        if (['<', '>'].includes(move)) {
          // left/right
          let boxCol = newCol;
          // follow offsetCol until we find a non-box
          while (['[', ']'].includes(mapData[newRow][boxCol])) {
            boxCol += offsetCol;
          }
          if (DEBUG) console.debug('move left/right', { col, newCol, boxCol });
          // if the non-box is empty
          if (mapData[newRow][boxCol] == '.') {
            // loop over newCol to boxCol and rewrite chars
            for (let i = 1; i < Math.abs(newCol - boxCol) + 1; i++) {
              // map i back to the correct column
              const updateCol = newCol + i * offsetCol;
              const updateChar = i % 2 == (offsetCol > 0 ? 1 : 0) ? '[' : ']';
              if (DEBUG) console.debug('move left/right', { i, offsetCol, newCol, updateCol, updateChar });
              // set char based on i mod 2
              mapData[newRow][updateCol] = updateChar;
            }
            // move
            mapData[row][col] = '.';
            row = newRow;
            col = newCol;
            mapData[row][col] = '@';
          }
        } else {
          // up down requires a recursive function
          // return an array of coords as a number[] so sort works if we need it
          const { boxCoords, success } = walkVertBoxes(mapData, newRow, newCol, offsetRow);
          if (DEBUG) console.debug({ success, boxCoords });
          if (success) {
            if (boxCoords.length == 0) throw new Error('boxCoords cannot be empty');

            // these can be out of order
            const sortedBoxCoords = boxCoords.toSorted((a, b) => (a[0] - b[0] || a[1] - b[1]) * offsetRow * -1);

            for (const [boxRow, boxCol] of sortedBoxCoords) {
              if (DEBUG) console.debug({ boxRow, boxCol });
              mapData[boxRow + offsetRow][boxCol] = mapData[boxRow][boxCol];
              mapData[boxRow][boxCol] = '.';
            }
            // move
            mapData[row][col] = '.';
            row = newRow;
            col = newCol;
            mapData[row][col] = '@';
          }
        }
        break;
      default:
        throw new Error(`invalid map char ${charAt} at ${newRow},${newCol}`);
    }
  }
  const gpsTotal = mapData.reduce(
    (rowAcc, row, rowIx) => (rowAcc += row.reduce((colAcc, char, colIx) => (colAcc += char == '[' ? rowIx * 100 + colIx : 0), 0)),
    0
  );
  console.log('part 2:', gpsTotal);
};

part1(blocks[0], moves);
part2(blocks[0], moves);
